core/src/main/java/io/grpc/internal/ClientCallImpl.java

```diff
@@ -83,6 +87,8 @@
       ScheduledExecutorService deadlineCancellationExecutor) {
     this.method = method;
     this.callExecutor = executor;
+    // Propagate the context from the thread which initiated the call to all callbacks.
+    this.context = Context.current();
```

Comment:
Right now I don't think there's enough evidence to choose one way or another. Given that I'm inclined to go with construction-time for the moment.